libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,354 @@
+	for (string const& sourceUnitName: m_fileReader.sourceCodes() | ranges::views::keys)
+		diagnosticsBySourceUnit[sourceUnitName] = Json::arrayValue;
+
+	for (shared_ptr<Error const> const& error: m_compilerStack.errors())
+	{
+		SourceLocation const* location = error->sourceLocation();
+		if (!location || !location->sourceName)
+			// LSP only has diagnostics applied to individual files.
+			continue;
+
+		Json::Value jsonDiag;
+		jsonDiag["source"] = "solc";
+		jsonDiag["severity"] = toDiagnosticSeverity(error->type());
+		jsonDiag["code"] = Json::UInt64{error->errorId().error};
+		string message = error->typeName() + ":";
+		if (string const* comment = error->comment())
+			message += " " + *comment;
+		jsonDiag["message"] = move(message);
+		jsonDiag["range"] = toRange(*location);
+
+		if (auto const* secondary = error->secondarySourceLocation())
+			for (auto&& [secondaryMessage, secondaryLocation]: secondary->infos)
+			{
+				Json::Value jsonRelated;
+				jsonRelated["message"] = secondaryMessage;
+				jsonRelated["location"] = toJson(secondaryLocation);
+				jsonDiag["relatedInformation"].append(jsonRelated);
+			}
+
+		diagnosticsBySourceUnit[*location->sourceName].append(jsonDiag);
+	}
+
+	for (string const& sourceUnitName: m_fileReader.sourceCodes() | ranges::views::keys)
+	{
+		Json::Value params;
+		params["uri"] = sourceUnitNameToClientPath(sourceUnitName);
+		params["diagnostics"] = move(diagnosticsBySourceUnit.at(sourceUnitName));
+		m_client.notify("textDocument/publishDiagnostics", move(params));
+	}
+}
+
+bool LanguageServer::run()
+{
+	while (!m_exitRequested && !m_client.closed())
+	{
+		optional<Json::Value> const jsonMessage = m_client.receive();
+		if (!jsonMessage)
+		{
+			m_client.error({}, ErrorCode::ParseError, "Error parsing JSONRPC request.");
+			continue;
+		}
+
+		try
+		{
+			string const methodName = (*jsonMessage)["method"].asString();
+			MessageID const id = (*jsonMessage)["id"];
+
+			if (auto handler = valueOrDefault(m_handlers, methodName))
+				handler(id, (*jsonMessage)["params"]);
+			else
+				m_client.error(id, ErrorCode::MethodNotFound, "Unknown method " + methodName);
+		}
+		catch (exception const& e)
+		{
+			m_client.error({}, ErrorCode::InternalError, "Unhandled exception: "s + e.what());
+		}
+	}
+	return m_shutdownRequested;
+}
+
+void LanguageServer::handleInitialize(MessageID _id, Json::Value const& _args)
+{
+	// The default of FileReader is to use `.`, but the path from where the LSP was started
+	// should not matter.
+	string rootPath("/");
+	if (Json::Value uri = _args["rootUri"])
+	{
+		rootPath = uri.asString();
+		if (!boost::starts_with(rootPath, "file://"))
+		{
+			m_client.error(_id, ErrorCode::InvalidParams, "rootUri only supports file URI scheme.");
+			return;
+		}
+		rootPath = rootPath.substr(7);
+	}
+	else if (Json::Value rootPath = _args["rootPath"])
+		rootPath = rootPath.asString();
+
+	m_fileReader.setBasePath(boost::filesystem::path(rootPath));
```

Comment:
`string` is implicitly convertible to `boost::filesystem::path`.
```suggestion
	m_fileReader.setBasePath(rootPath);
```
Maybe it would be better if `rootPath` was `boost::filesystem::path`?